Drop Python 3.6

JonathanPlasse opened the pull request:
Close #141

empicano commented on asyncio_mqtt/client.py:
```diff
@@ -1,52 +1,37 @@
     Awaitable,
     Callable,
     Coroutine,
     Dict,
```
I'm not very experienced with typing in Python. Is this on purpose that you're leaving some `typing.Dict`, `typing.Tuple`, and `typing.Union` in the code, instead of changing everything to `dict`, `tuple` and `|`?

JonathanPlasse replied:
Yes, the typing used as type annotations can be `dict`, `tuple`, and `|` as they are not executed with `from __future__ import annotations`, but types that are assigned to variables are executed so they must remain as they are.

empicano replied:
And the non-`typing` versions `dict`, `tuple`, and `|` mustn't be executed because they are not valid python<3.10 code, is that it?

Neat 😎

JonathanPlasse replied:
Yes

frederikaalund commented:
LGTM. :+1:

I like the use of the new-style `dict`, `type`, `list`, `|` typings combined with `from __future__ import annotations`. Good way to use new features in a backwards-compatible way.

In a future commit, let's add some type aliases for convenience. E.g.:
```python
Topic = (
    str
    | tuple[str, mqtt.SubscribeOptions]
    | list[tuple[str, mqtt.SubscribeOptions]]
    | list[tuple[str, int]]
)
```

Even if we only use the alias in a single place, it's a little verbose to read through as part of a function parameter list.

In any case, good job with this PR. :+1: Great to see that you also fixed the tests and added the "supported python versions" badge to the documentation (and setup).

frederikaalund reviewed (approved):
Approved. :)

I see that you still some recent commits in this PR/branch. Do you want to add some other finishing touches before the merge or can I go ahead and merge it as-is?

JonathanPlasse commented:
> LGTM. +1
> 
> I like the use of the new-style `dict`, `type`, `list`, `|` typings combined with `from __future__ import annotations`. Good way to use new features in a backwards-compatible way.
> 
> In a future commit, let's add some type aliases for convenience. E.g.:
> 
> ```python
> Topic = (
>     str
>     | tuple[str, mqtt.SubscribeOptions]
>     | list[tuple[str, mqtt.SubscribeOptions]]
>     | list[tuple[str, int]]
> )
> ```
> 
> Even if we only use the alias in a single place, it's a little verbose to read through as part of a function parameter list.
> 
> In any case, good job with this PR. +1 Great to see that you also fixed the tests and added the "supported python versions" badge to the documentation (and setup).

The user would lose the details and only see the type alias. Is it what we want.

JonathanPlasse commented:
It is ready to be merged.
Which options do you use to merge?
- Create a merge commit
- Rebase and merge

For me, the cleaner option for a better history is rebasing.

frederikaalund commented:
*Rebase and merge*: If the commits are "clean" (e.g., not "TODO: Fix this" or "WIP: still developing on X").
*Squash and merge*: Otherwise.
*Create a merge commit*: Never.

For this PR, I think that we are in the first category: *Rebase and merge*. :+1:

JonathanPlasse commented:
I removed the option *Create a merge commit* to avoid errors.
